docs/site/Dynamic-models-repositories-controllers.md

```diff
@@ -0,0 +1,218 @@
+      mysql: {
+        columnName: 'id',
+        dataType: 'int',
+        dataLength: null,
+        dataPrecision: 10,
+        dataScale: 0,
+        nullable: 'N',
+      },
+    },
+    title: {
+      type: 'String',
+      required: false,
+      length: 512,
+      precision: null,
+      scale: null,
+      mysql: {
+        columnName: 'title',
+        dataType: 'varchar',
+        dataLength: 512,
+        dataPrecision: null,
+        dataScale: null,
+        nullable: 'Y',
+      },
+    }
+  },
+  settings: {
+    idInjection: false,
+    mysql: {schema: 'test', table: 'BookStore'},
+  },
+});
+```
+
+### Defining a Model
+
+A LoopBack model class is created by passing a `ModelDefinition` object to
+`@loopback/repository`'s helper function `defineModelClass()`. It expects a
+base model to extend (typically `Model` or `Entity`), folowed by the model
+definition object. In this case it will be `Entity`.
+
+```ts
+const BookModel = defineModelClass(Entity, bookDef);
+```
+
+In case you need to use an existing Model as the base class, you should use the
+`DynamicModelCtor` class, instead of the `defineModelClass()` function.
+
+```ts
+import DynamicModelCtor from '@loopback/repository';
+// Assuming User is a pre-existing Model class in the app
+const StudentModel = new DynamicModelCtor(User, studentDef);
+```
+
+For details about `ModelDefinition`, `defineModelClass`, `Model`, `Entity`, and
+`DynamicModelCtor` refer to the
+[@loopback/repository API documentation](https://loopback.io/doc/en/lb4/apidocs.repository.html).
+
+### Defining a Datasource
+
+Before a repository for this model can be set up, a datasource should be ready.
+If required, set one up dynamically by creating an instance of
+`juggler.DataSource`. It requires a name for the datasource, the connector, and
+a connection url as shown below.
+
+```ts
+const dsName = 'bookstore-ds';
+const bookDs = new juggler.DataSource({
+  name: dsName,
+  connector: require('loopback-connector-mongodb'),
+  url: 'mongodb://sysop:moon@localhost',
+});
+await bookDs.connect();
+app.dataSource(bookDs, dsName);
+```
+
+Connect the datsource by calling its `.connect()` method and attach it to the
+app using `app.dataSource()` method. Note, this method will be available only
+on `RepositoryMixin` apps.
+
+### Defining a Repository
+
+Now that a datasource is ready, a repository can be created for `BookModel`.
+
+The `@loopback/rest-crud` package exports a helper function,
+`defineCrudRepositoryClass`, this function creates a
+`defineCrudRepositoryClass`-based entity CRUD repository class for the given
+model.
+
+```ts
+const BookRepository = defineCrudRepositoryClass(BookModel);
+```
+
+If you want to customize the repository, you can create a copy of
+`defineCrudRepositoryClass`'s
+[implementation](https://github.com/strongloop/loopback-next/blob/00917f5a06ea8a51e1f452f228a6b0b7314809be/packages/rest-crud/src/repository-builder.ts#L25-L47)
+and modify it according to your requirements.
```

Comment:
I don't think this is the best advice. `defineCrudRepositoryClass` is just a thin wrapper for `defineRepositoryClass`, which is in turn a small function creating a named repository class.

There are several other options how users can customize their repositories.

1. Put the custom behavior into a repository base class. Use that base class as the base to inherit from.

    ```ts
    class MyRepoBase<
      E extends Entity,
      IdType,
      Relations extends object
    > extends DefaultCrudRepository<E, IdType, Relations> {
      // put your custom code here
    }

    const BookRepositoryClass = defineRepositoryClass<
      typeof BookModel,
      MyRepoBase<BookModel, typeof BookModel.prototype.id, BookRelations>
    >(BookModel, MyRepoBase);
    ```

2. Use Repository Mixins to implement custom behavior and then apply the mixins when creating the repository class.

   Assuming we already have [FindByTitleRepositoryMixin from our docs](https://loopback.io/doc/en/lb4/migration-models-mixins.html#defining-a-repository-mixin-class-factory-function):

    ```ts
    const BookRepositoryClass = FindByTitleRepositoryMixin(
      defineCrudRepositoryClass<
        BookModel,
        typeof BookModel.prototype.id,
        BookRelations
      >(BookModel),
    );
    ```

    The code above is simple, but may define a class constructor with incorrect name, because we create `BookRepository` first and then apply the mixin second. Here is a bullet-proof approach that is unfortunately more complex:

    ```ts
    const BookRepositoryClass = defineRepositoryClass<
      typeof BookModel,
      DefaultCrudRepository<
        BookModel,
        typeof BookModel.prototype.id,
        BookRelations
      > &
        FindByTitle<BookModel>
    >(BookModel, FindByTitleRepositoryMixin(DefaultCrudRepository));
    ```

Please rework the paragraph about repository customization to show both Repository Base Class and Repository Mixin options.

Reply:
Loosely related: https://github.com/strongloop/loopback-next/issues/5565 and [How to reuse LoopBack repository code](https://youtu.be/hRb70_JcNWI)